### ISSUE TITLE
Ability to specify versions for extension dependencies

### ISSUE THREAD

**Issue by Piedone** [OrchardCMS/Orchard#5703]
Like how NuGet does it: you could e.g. say that my module needs ModuleB, but with version >= 1.2.

**Comment by bleroy** [comment]
Npm does that too, and that was considered when the initial design was made. It comes with some additional complexity and doesn't necessarily solve the problem: if version 1.3 happens to be breaking you in the future, you have no way of knowing in advance. And if you are too restrictive in what version you declare, you might just force yourself into an unnecessary update in the future. On the other hand, if version 1.3 breaks you in the future, you're screwed no matter what.
I prefer an evidence-based approach where we give ourselves the telemetry data, and crowdsourced information about what really works.
This is definitely not an easy problem.

**Comment by Piedone** [comment]
This certainly isn't an easy problem. I'm more like thinking about hints rather than enforcing rules as it is with dependency declarations (and e.g. display a warning on the Feature screen when a module says it needs a different version of another module).

This won't prevent you from experimenting or using a module even when its dependency version declaration is out of date but it can at least give you hints: I'd definitely keep the version declaration optional so when an extension author does specify it it has a good reason.

**Comment by bleroy** [comment]
Yes, that sounds reasonable. Maybe bring up the issue at the next meeting?

**Comment by Piedone** [comment]
Will do.

**Comment by sebastienros** [comment]
Notes: Impact on the gallery page in the admin, should we show there are incompatibilities with the currently installed modules. Should we allow to download old (compatible) versions.
should the gallery show the dependencies and the versions.
